File: server/routers/papers.py
"""论文搜集路由 — 代理到 SeekVerse API"""
import json
from pathlib import Path

import httpx
from fastapi import APIRouter, Body, Request, Response
from fastapi.responses import StreamingResponse
from server.config import SEEKVERSE_BASE

router = APIRouter(prefix="/api/papers", tags=["papers"])

_client = httpx.AsyncClient(base_url=SEEKVERSE_BASE, timeout=30.0)

# ========== 本地收藏/置顶存储 ==========

_FAV_PATH = Path("data/paper_favorites.json")


def _load_favs() -> dict:
    """加载本地收藏/置顶数据。"""
    if _FAV_PATH.exists():
        return json.loads(_FAV_PATH.read_text(encoding="utf-8"))
    return {}


def _save_favs(favs: dict) -> None:
    """保存收藏/置顶数据。"""
    _FAV_PATH.parent.mkdir(parents=True, exist_ok=True)
    _FAV_PATH.write_text(json.dumps(favs, ensure_ascii=False, indent=2), encoding="utf-8")


def _enrich_local(paper_dict: dict) -> dict:
    """为论文 dict 附加本地 starred/pinned 状态。"""
    favs = _load_favs()
    pid = paper_dict.get("id", "")
    fav = favs.get(pid, {})
    paper_dict["starred"] = fav.get("starred", False)
    paper_dict["pinned"] = fav.get("pinned", False)
    return paper_dict


@router.get("")
async def list_papers(request: Request):
    """获取论文列表 — 代理 SeekVerse，附加本地收藏/置顶状态"""
    params = dict(request.query_params)
    resp = await _client.get("/api/papers", params=params)
    data = resp.json()
    for p in data.get("papers", []):
        _enrich_local(p)
    return data


@router.get("/stats/summary")
async def get_stats():
    """获取论文统计 — 代理 SeekVerse"""
    resp = await _client.get("/api/papers/stats/summary")
    return resp.json()


@router.post("/fetch")
async def fetch_paper(url: str = Body(..., embed=True), force: bool = Body(False, embed=True)):
    """通过 URL 拉取论文 — 代理 SeekVerse"""
    resp = await _client.post("/api/papers/fetch", json={"url": url, "force": force})
    return resp.json()


@router.get("/{paper_id}")
async def get_paper_detail(paper_id: str):
    """获取论文详情 — 代理 SeekVerse，附加本地收藏/置顶状态"""
    resp = await _client.get(f"/api/papers/{paper_id}")
    data = resp.json()
    return _enrich_local(data)


@router.get("/{paper_id}/note")
async def get_paper_note(paper_id: str):
    """获取论文笔记 — 代理 SeekVerse"""
    resp = await _client.get(f"/api/papers/{paper_id}/note")
    return resp.json()


@router.put("/{paper_id}/note")
async def save_paper_note(paper_id: str, content: str = Body(..., embed=True)):
    """保存论文笔记 — 代理 SeekVerse"""
    resp = await _client.put(f"/api/papers/{paper_id}/note", json={"content": content})
    return resp.json()


@router.put("/{paper_id}/blog")
async def set_blog_url(paper_id: str, blog_url: str = Body(..., embed=True)):
    """设置论文 blog 链接 — 代理 SeekVerse"""
    resp = await _client.put(f"/api/papers/{paper_id}/blog", json={"blog_url": blog_url})
    return resp.json()


@router.post("/{paper_id}/summarize")
async def summarize_paper(paper_id: str):
    """触发论文摘要生成 — 代理 SeekVerse"""
    resp = await _client.post(f"/api/papers/{paper_id}/summarize")
    return resp.json()


# ========== 收藏 / 置顶 ==========

@router.put("/{paper_id}/star")
async def toggle_star(paper_id: str, starred: bool = Body(..., embed=True)):
    """收藏 / 取消收藏论文。"""
    favs = _load_favs()
    if paper_id not in favs:
        favs[paper_id] = {}
    favs[paper_id]["starred"] = starred
    _save_favs(favs)
    return {"status": "ok", "starred": starred}


@router.put("/{paper_id}/pin")
async def toggle_pin(paper_id: str, pinned: bool = Body(..., embed=True)):
    """置顶 / 取消置顶论文。"""
    favs = _load_favs()
    if paper_id not in favs:
        favs[paper_id] = {}
    favs[paper_id]["pinned"] = pinned
    _save_favs(favs)
    return {"status": "ok", "pinned": pinned}


# ========== 缩略图代理 ==========

@router.get("/thumbnails/{arxiv_id}")
async def get_thumbnail(arxiv_id: str):
    """获取论文缩略图 — 代理 SeekVerse。

    SeekVerse 在缩略图不存在时返回 JSON 错误（状态码 200），
    需要检查 content-type 判断是否为真实图片。
    """
    # 尝试 webp
    resp = await _client.get(f"/thumbnails/{arxiv_id}.webp")
    ct = resp.headers.get("content-type", "")
    if resp.status_code == 200 and "image" in ct:
        return Response(content=resp.content, media_type="image/webp")

    # 尝试 png
    resp = await _client.get(f"/thumbnails/{arxiv_id}.png")
    ct = resp.headers.get("content-type", "")
    if resp.status_code == 200 and "image" in ct:
        return Response(content=resp.content, media_type="image/png")

    return Response(content=b"", media_type="image/png", status_code=404)


# ========== Digest 代理 ==========

@router.get("/digests/daily")
async def get_daily_digest():
    """获取今日论文 digest — 代理 SeekVerse"""
    resp = await _client.get("/api/digests/daily")
    return resp.json()


@router.post("/digests/generate")
async def generate_digest(period: str = "今日", limit: int = 100):
    """生成论文 digest — 代理 SeekVerse"""
    resp = await _client.post("/api/digests/generate", params={"period": period, "limit": limit})
    return resp.json()

File: web/src/api/papers.js
/**
 * 论文 API 封装
 *
 * 开发环境：通过 /api 代理到后端 FastAPI
 * 生产环境（GitHub Pages）：从静态 JSON 加载，收藏/置顶/笔记用 localStorage
 */
import request from './request'

const isProd = import.meta.env.PROD

// ========== localStorage 工具 ==========

const LS_KEY = 'par_favorites'
const LS_NOTE_KEY = 'par_notes'

function loadFavs() {
  try {
    return JSON.parse(localStorage.getItem(LS_KEY) || '{}')
  } catch {
    return {}
  }
}

function saveFavs(favs) {
  localStorage.setItem(LS_KEY, JSON.stringify(favs))
}

function loadNotes() {
  try {
    return JSON.parse(localStorage.getItem(LS_NOTE_KEY) || '{}')
  } catch {
    return {}
  }
}

function saveNotes(notes) {
  localStorage.setItem(LS_NOTE_KEY, JSON.stringify(notes))
}

// ========== 静态数据加载 ==========

let _staticPapers = null

async function loadStaticPapers() {
  if (_staticPapers) return _staticPapers
  const base = import.meta.env.BASE_URL || '/'
  const resp = await fetch(`${base}data/papers.json`)
  const data = await resp.json()

  // 注入 localStorage 中的收藏/置顶状态
  const favs = loadFavs()
  const notes = loadNotes()
  data.papers.forEach(p => {
    const fav = favs[p.id] || {}
    p.starred = fav.starred || false
    p.pinned = fav.pinned || false
    p.note = notes[p.id] || ''
    p.has_note = !!p.note
  })

  _staticPapers = data
  return data
}

// ========== API 导出 ==========

// 论文列表
export const getPaperList = async (params = {}) => {
  if (isProd) {
    const data = await loadStaticPapers()
    let papers = data.papers

    // 客户端筛选
    if (params.category) {
      papers = papers.filter(p => (p.categories || []).includes(params.category))
    }
    if (params.source) {
      papers = papers.filter(p => p.source === params.source)
    }

    const limit = params.limit || 100
    const offset = params.offset || 0
    return {
      total: papers.length,
      papers: papers.slice(offset, offset + limit),
    }
  }
  return request.get('/papers', { params })
}

// 论文统计
export const getPaperStats = async () => {
  if (isProd) {
    const data = await loadStaticPapers()
    const by_source = {}
    const by_category = {}
    data.papers.forEach(p => {
      by_source[p.source] = (by_source[p.source] || 0) + 1
      ;(p.categories || []).forEach(c => {
        by_category[c] = (by_category[c] || 0) + 1
      })
    })
    return { total: data.papers.length, by_source, by_category }
  }
  return request.get('/papers/stats/summary')
}

// 论文详情
export const getPaperDetail = async (id) => {
  if (isProd) {
    const data = await loadStaticPapers()
    return data.papers.find(p => p.id === id) || { error: 'not found' }
  }
  return request.get(`/papers/${id}`)
}

// 论文笔记
export const getPaperNote = async (id) => {
  if (isProd) {
    const notes = loadNotes()
    return { content: notes[id] || '' }
  }
  return request.get(`/papers/${id}/note`)
}

export const savePaperNote = async (id, content) => {
  if (isProd) {
    const notes = loadNotes()
    notes[id] = content
    saveNotes(notes)
    // 更新缓存
    if (_staticPapers) {
      const p = _staticPapers.papers.find(p => p.id === id)
      if (p) { p.note = content; p.has_note = !!content }
    }
    return { status: 'ok' }
  }
  return request.put(`/papers/${id}/note`, { content })
}

// Blog 链接
export const setPaperBlog = async (id, blogUrl) => {
  if (isProd) {
    if (_staticPapers) {
      const p = _staticPapers.papers.find(p => p.id === id)
      if (p) p.blog_url = blogUrl
    }
    return { status: 'ok' }
  }
  return request.put(`/papers/${id}/blog`, { blog_url: blogUrl })
}

// 通过 URL 拉取论文（静态部署不支持）
export const fetchPaper = async (url, force = false) => {
  if (isProd) return { error: '静态部署不支持添加论文' }
  return request.post('/papers/fetch', { url, force })
}

// 触发摘要生成（静态部署不支持）
export const summarizePaper = async (id) => {
  if (isProd) return { status: 'not_implemented' }
  return request.post(`/papers/${id}/summarize`)
}

// 收藏 / 取消收藏
export const starPaper = async (id, starred) => {
  if (isProd) {
    const favs = loadFavs()
    if (!favs[id]) favs[id] = {}
    favs[id].starred = starred
    saveFavs(favs)
    if (_staticPapers) {
      const p = _staticPapers.papers.find(p => p.id === id)
      if (p) p.starred = starred
    }
    return { status: 'ok', starred }
  }
  return request.put(`/papers/${id}/star`, { starred })
}

// 置顶 / 取消置顶
export const pinPaper = async (id, pinned) => {
  if (isProd) {
    const favs = loadFavs()
    if (!favs[id]) favs[id] = {}
    favs[id].pinned = pinned
    saveFavs(favs)
    if (_staticPapers) {
      const p = _staticPapers.papers.find(p => p.id === id)
      if (p) p.pinned = pinned
    }
    return { status: 'ok', pinned }
  }
  return request.put(`/papers/${id}/pin`, { pinned })
}

// 缩略图 URL
export const getThumbnailUrl = (arxivId) => {
  if (isProd) return null  // 静态部署无缩略图
  return `/api/papers/thumbnails/${arxivId}`
}

// 论文 digest（静态部署不支持）
export const getDailyDigest = async () => {
  if (isProd) return { digest: null }
  return request.get('/papers/digests/daily')
}

export const generateDigest = async (period = '今日', limit = 100) => {
  if (isProd) return { error: '静态部署不支持' }
  return request.post('/papers/digests/generate', null, { params: { period, limit } })
}

// 保留旧接口兼容
export const getPaperCategories = () => Promise.resolve(['action_recognition', 'pet_action_recognition', 'pose_estimation', 'skeleton_action_recognition', 'survey', 'temporal_action_detection', 'video_foundation_model'])
export const getPaperSources = async () => {
  if (isProd) return { sources: ['arxiv', 'manual'] }
  return request.get('/papers/sources')
}
export const updatePaperSource = async (id, data) => {
  if (isProd) return { status: 'ok' }
  return request.put(`/papers/sources/${id}`, data)
}
import request from './request'

// 论文列表（代理 SeekVerse）
export const getPaperList = (params) => request.get('/papers', { params })

// 论文统计
export const getPaperStats = () => request.get('/papers/stats/summary')

// 论文详情
export const getPaperDetail = (id) => request.get(`/papers/${id}`)

// 论文笔记
export const getPaperNote = (id) => request.get(`/papers/${id}/note`)
export const savePaperNote = (id, content) => request.put(`/papers/${id}/note`, { content })

// Blog 链接
export const setPaperBlog = (id, blogUrl) => request.put(`/papers/${id}/blog`, { blog_url: blogUrl })

// 通过 URL 拉取论文
export const fetchPaper = (url, force = false) => request.post('/papers/fetch', { url, force })

// 触发摘要生成
export const summarizePaper = (id) => request.post(`/papers/${id}/summarize`)

// 收藏 / 取消收藏
export const starPaper = (id, starred) => request.put(`/papers/${id}/star`, { starred })

// 置顶 / 取消置顶
export const pinPaper = (id, pinned) => request.put(`/papers/${id}/pin`, { pinned })

// 缩略图 URL
export const getThumbnailUrl = (arxivId) => `/api/papers/thumbnails/${arxivId}`

// 论文 digest
export const getDailyDigest = () => request.get('/papers/digests/daily')
export const generateDigest = (period = '今日', limit = 100) => request.post('/papers/digests/generate', null, { params: { period, limit } })

// 保留旧接口兼容
export const getPaperCategories = () => Promise.resolve(['autoregressive', 'diffusion', 'image_compression', 'visual_tokenizer'])
export const getPaperSources = () => request.get('/papers/sources')
export const updatePaperSource = (id, data) => request.put(`/papers/sources/${id}`, data)
